File: algorithm/131-PalindromePartitioning/PalindromePartitioning.cpp
// Source: https://leetcode.com/problems/palindrome-partitioning/description/
// Author: Diego Lee
// Date: 2018-10-10
//
// Description:
// Given a string s, partition s such that every substring of the partition is a palindrome.
//
// Return all possible palindrome partitioning of s.
//
// Example:
//
// Input: "aab"
// Output:
// [
//   ["aa","b"],
//   ["a","a","b"]
// ]

#include <iostream>
#include <string>
#include <vector>
#include <cassert>

class Solution {
    public:
        std::vector<std::vector<std::string>> partition(std::string s) {
            std::vector<std::vector<std::string>> result;
            for (int i = 1, len = s.size(); i <= len; ++i) {
                std::string part = s.substr(0, i);
                if (isPalindrome(part)) {
                    if (i == len) {
                        result.push_back(std::vector<std::string>({part}));
                    }
                    else {
                        std::vector<std::vector<std::string>> curr = partition(s.substr(i, len - i));
                        if (!curr.empty()) {
                            for (size_t j = 0; j < curr.size(); ++j) {
                                curr[j].insert(curr[j].begin(), part);
                                result.push_back(curr[j]);
                            }
                        }
                    }
                }
            }
            return result;
        }

        bool isPalindrome(const std::string& s) {
            int left = 0;
            int right = s.size() - 1;
            while (left <= right) {
                if (s[left++] != s[right--]) {
                    return false;
                }
            }
            return true;
        }
};

void testPartition() {
    Solution s;
    std::vector<std::vector<std::string>> result = s.partition("aab");
    assert(result.size() == 2);
    assert(result[0] == std::vector<std::string>({"a", "a", "b"}));
    assert(result[1] == std::vector<std::string>({"aa", "b"}));
}

int main() {
    testPartition();
}

File: algorithm/131-PalindromePartitioning/go/PalindromePartitioning.go
package solution

func partition(s string) [][]string {
    if len(s) == 0 {
        return [][]string{}
    }

    result := [][]string{}
    length := len(s)
    for i := 1; i <= length; i++ {
        part := s[:i]
        if isPalindrome(part) {
            if i == length {
                result = append(result, []string{part})
            } else {
                curr := partition(s[i:])
                if len(curr) != 0 {
                    for j, _ := range curr {
                        curr[j] = append(curr[j][:0], append([]string{part}, curr[j][0:]...)...)
                        result = append(result, curr[j])
                    }
                }
            }
        }
    }
    return result
}

func isPalindrome(s string) bool {
    left := 0
    right := len(s) - 1
    for left <= right {
        if s[left] != s[right] {
            return false
        }
        left += 1
        right -= 1
    }
    return true
}
